MobiusCore/Source/EffectHandlers/EffectHandler.swift

```diff
@@ -17,16 +17,44 @@
 // specific language governing permissions and limitations
 // under the License.
 
-// Primitive EffectHandler
-public struct EffectHandler<Effect, Event> {
-    public let handle: (Effect, @escaping Consumer<Event>) -> Void
-    public let disposable: Disposable
+public protocol EffectHandler {
+    associatedtype Effect
+    associatedtype Event
+
+    /// Handle an `Effect`.
+    /// To output events, call `response.send`
+    /// When you are done handling `input`, be sure to call `response.end()` to prevent memory leaks.
+    ///
+    /// Return a `Disposable` which tears down any resources that is being used by this effect handler.
+    func handle(
+        _ input: Effect,
+        _ response: Response<Event>
+    ) -> Disposable
+}
+
+public class Response<T> {
+    public let send: (T) -> Void
+    public let end: () -> Void
 
     public init(
-        handle: @escaping (Effect, @escaping Consumer<Event>) -> Void,
-        disposable: Disposable
+        onSend: @escaping (T) -> Void,
+        onEnd: @escaping () -> Void
     ) {
-        self.handle = handle
-        self.disposable = disposable
+        self.send = onSend
+        self.end = onEnd
+    }
+}
```

Comment:
I would much prefer it if `Response` called `end` in `deinit`. If that’s not acceptable, it should `fatalError` if `end` hasn’t been called. I don’t think silently leaking responses would be acceptable.